db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/PredicateList.java

```diff
@@ -1111,15 +1116,28 @@ private void orderUsefulPredicates(Optimizable optTable,
         ArrayList<Predicate> predsForNewInList = null;
     
         if (inlistQualified) {
-          // Only combine multiple IN lists if not using Spark.
-          // Adding extra RDDs and unioning them together hinders performance.
-          if (inlistPreds.size() > 1) {
+          // In DB2 mode, probe keys can be overlapping with eachother.
+          // On HBase the overlapping keys are merged together, but on
+          // mem they aren't causing duplicate rows.  So here we avoid
```

Comment:
Can we fix this on Mem by merging those keys? I dislike introducing knowledge of the platform in the engine itself.

Reply:
It's possible, but may take a bit of effort for not much performance benefit.  It's not actually merging that could be done on mem, but expansion, since each probe value is applied independently.  For example, if we have t1 (a varchar(101), b int);  create index ix1 on t1 (a, b);  select * from t1 where a in ('a', 'b', 'c') and b in (1,2,3);  Each probe value for column a would have to be chopped up into 100 probe values to cover each possible number of trailing spaces.  For this case that would be 900 total probe values.  It would likely be more efficient to just use keys from column a resulting in 3 start/stop keys, and apply the column b predicate as a filter.  There really is a difference in how mem does probing vs. HBase.  HBase does a straight scan through rows in a single pass, while mem does actual probing where each probe is done independently.  If we don't want the optimizer to be aware of the platform, we would either have to limit which predicates we use on HBase to match mem, or we could change mem to do a straight scan through in a single thread.  Keep in mind that this is only for DB2 compatibility mode and only mem platform that customers never see, which we are talking about optimizing.  I would argue that the implementation effort compared to possible benefit isn't worth it.

Reply:
After a discussion with @msirek we thought it better to open a JIRA to address this at a later time and keep this optimization here for now.